### PR TITLE
Remove empty SuccessfulPlutusScriptsEvent events

### LINKED ISSUE #2859: SuccessfulPlutusScriptsEvent emitted too often
The `SuccessfulPlutusScriptsEvent` event is emitted even when a transaction contains _no_ Plutus scripts:

https://github.com/input-output-hk/cardano-ledger/blob/0b0d066bf35f4062545bfaeeaf2cca9a1a87a2c3/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs#L243

The path to `Passes ps` contains the case that `ps` is empty.

We should not emit this event when `ps` is empty. We should use `NonEmpty` instead of a List inside `SuccessfulPlutusScriptsEvent` (this also makes sense in the branch where a failed Plutus script exists).

### PR DESCRIPTION
**Pull request by JaredCorduan**
The `SuccessfulPlutusScriptsEvent` event should only be emitted when the list of successful Plutus scripts is nonempty. I've made `SuccessfulPlutusScriptsEvent` use `NonEmpty` instead of list.

Additionally, the `SuccessfulPlutusScriptsEvent` events were not being emitted at all in the Babbage era, in the no-failures path.

resolves #2859

### CODE REVIEW

**Review by teodanciu — APPROVED** [review]
Looks good to me.

I have a vague feeling  that there is some abstraction to express `maybe (pure ()) f someMaybe`, but I don't know which and playing a bit with it lead to no results! It was more a curiosity, this is perfectly fine  :+1:

**Comment by JaredCorduan** [comment]
> Looks good to me.
> 
> I have a vague feeling that there is some abstraction to express `maybe (pure ()) f someMaybe`, but I don't know which and playing a bit with it lead to no results! It was more a curiosity, this is perfectly fine +1

Thanks @teodanciu ! I'm happy to leave the PR up for a while and see if anyone else knows a better way to write it. I had the same vague feeling...

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs [hunk 1]
```diff
@@ -240,7 +240,7 @@ scriptsValidateTransition = do
         ValidationTagMismatch
           (getField @"isValid" tx)
           (FailedUnexpectedly (scriptFailuresToPredicateFailure fs))
-    Passes ps -> tellEvent (SuccessfulPlutusScriptsEvent ps)
+    Passes ps -> maybe (pure ()) (tellEvent . SuccessfulPlutusScriptsEvent) (nonEmpty ps)
```
Comment (lehins) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs]:
```suggestion
    Passes ps -> mapM_ (tellEvent . SuccessfulPlutusScriptsEvent) (nonEmpty ps)
```

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs [hunk 2]
```diff
@@ -270,7 +270,7 @@ scriptsNotValidateTransition = do
       failBecause $
         ValidationTagMismatch (getField @"isValid" tx) PassedUnexpectedly
     Fails ps fs -> do
-      tellEvent (SuccessfulPlutusScriptsEvent ps)
+      maybe (pure ()) (tellEvent . SuccessfulPlutusScriptsEvent) (nonEmpty ps)
```
Comment (lehins) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs]:
```suggestion
      mapM_ (tellEvent . SuccessfulPlutusScriptsEvent) (nonEmpty ps)
```

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxos.hs [hunk 3]
```diff
@@ -208,7 +209,7 @@ scriptsNo = do
         when2Phase $ case evalScripts @era (getField @"_protocolVersion" pp) tx sLst of
           Passes _ -> failBecause $ ValidationTagMismatch (getField @"isValid" tx) PassedUnexpectedly
           Fails ps fs -> do
-            tellEvent (SuccessfulPlutusScriptsEvent ps)
+            maybe (pure ()) (tellEvent . SuccessfulPlutusScriptsEvent) (nonEmpty ps)
```
Comment (lehins) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxos.hs]:
```suggestion
            mapM_ (tellEvent . SuccessfulPlutusScriptsEvent) (nonEmpty ps)
```

**Review by lehins — APPROVED** [review]
@teodanciu and @JaredCorduan abstraction you are looking for called `mapM_`/`traverse_` :wink: 

This looks great. It always makes me happy when a switch to NonEmpty happens :smile:

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxos.hs [hunk 4]
```diff
@@ -175,7 +176,7 @@ scriptsYes = do
                 ValidationTagMismatch
                   (getField @"isValid" tx)
                   (FailedUnexpectedly (scriptFailuresToPredicateFailure fs))
-            Passes _ -> pure ()
+            Passes ps -> maybe (pure ()) (tellEvent . SuccessfulPlutusScriptsEvent) (nonEmpty ps)
```
Comment (JaredCorduan) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxos.hs]: I'll snag this one too! :)

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs (hunk 1 above)
Comment (Soupstraw) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs]: Dammit, I actually wanted to suggest the same, but I forgot to submit the review after I made the suggestions :laughing:

**Review by Soupstraw — APPROVED** [review]

**Pull request merged**